modules/renderer/features.js

```diff
@@ -94,6 +96,44 @@ export function rendererFeatures(context) {
         };
     }
 
+    defineRule('date_range', function isWithinRange(tags) {
+      // dates a are a mix of yyyy-mm-dd yyyy-mm and yyyy, so pad them out
+      // then use DecimalDate JS https://github.com/OpenHistoricalMap/decimaldate-javascript/ to properly handle negative BCE dates
+      var dateToNumber = function(date) {
+        if (! date) return NaN;
+
+        let paddeddate;
+        if (date.match(/^\-?\d\d\d\d\-\d\d\-\d\d$/)) {
+          paddeddate = date; // perfect yyyy-mm-dd
+        } else if (date.match(/^\-?\d\d\d\d\-\d\d$/)) {
+          const ym = date.match(/^\-?(\d\d\d\d)\-(\d\d)$/);
+          const d = (new Date(parseInt(ym[1]), parseInt(ym[2]), 0)).getDate();
+          paddeddate = date + `-${String(d).padStart(2, '0')}`; // yyyy-mm, add -dd
+        } else if (date.match(/^\-?\d\d\d\d$/)) {
+          paddeddate = date + '-12-31'; // yyyy, add -mm-dd
+        } else if (date.match(/^\-?\d{1,3}$/)) {
+          paddeddate = date.padStart(4, '0') + '-12-31';
+        }
+
+        return decimaldate.iso2dec(paddeddate);
+      };
+
+      // entity's start & end date
+      // filtering date range from the on-screen controls
+      const entityRange = {
+        'start_date': dateToNumber(tags.start_date),
+        'end_date': dateToNumber(tags.end_date)
+      };
+      const selectedRange = {
+        'start_date': dateToNumber(context.features().dateRange[0]),
+        'end_date': dateToNumber(context.features().dateRange[1])
+      };
```

Comment:
The new validator in #138 handles overflow and padding too. It adds another util.js method that could come in handy here:

```js
const selectedRange = {
  'start_date': utilNormalizeDateString(context.features().dateRange[0]).value,
  'end_date': utilNormalizeDateString(context.features().dateRange[1]).value
};
```

Reply:
I have changed to use `utilNormalizeDateString()` and `utilDatesOverlap()`